File: Compiler/src/util/DataType.java
package util;

// FIXME: заглушка. Типы данных используются только для валидации, поэтому пока достаточно ENUM
// TODO?: тип данных ENUM?
/**
 * Типы данных
 */
public enum DataType {

    /**
     * Переменная дерева мысли
     */
    DECISION_TREE_VAR,

    /**
     * Класс
     */
    CLASS,

    /**
     * Объект
     */
    OBJECT,

    /**
     * Свойство
     */
    PROPERTY,

    /**
     * Отношение
     */
    RELATIONSHIP,

    /**
     * Строка
     */
    STRING,

    /**
     * Булево значение
     */
    BOOLEAN,

    /**
     * Целое число
     */
    INTEGER,

    /**
     * Дробное число
     */
    DOUBLE,

    /**
     * Результат сравнения
     */
    COMPARISON_RESULT,

    /**
     * Литерал (Строка | Булево значение | Целое число | Дробное число)
     */
    LITERAL;

    /**
     * Может ли один тип быть преобразован в другой
     * @param from Тип, который преобразовываем
     * @param to Тип, в который преобразовываем
     * @return Может ли один тип быть преобразован в другой
     */
    public static boolean canCast(DataType from, DataType to) {
        return from == DECISION_TREE_VAR && to == OBJECT ||
                from == STRING && to == LITERAL ||
                from == BOOLEAN && to == LITERAL ||
                from == INTEGER && to == LITERAL ||
                from == DOUBLE && to == LITERAL;
    }

    /**
     * Может ли этот тип быть преобразован в другой
     * @param to Тип, в который преобразовываем
     * @return Может ли этот тип быть преобразован в другой
     */
    public boolean canCast(DataType to) {
        return this == DECISION_TREE_VAR && to == OBJECT ||
                this == STRING && to == LITERAL ||
                this == BOOLEAN && to == LITERAL ||
                this == INTEGER && to == LITERAL ||
                this == DOUBLE && to == LITERAL;
    }
}

File: Compiler/src/util/DataType.kt
package util

/**
 * Тип данных
 */
sealed class DataType {

    /**
     * Переменная дерева мысли
     */
    object DecisionTreeVar : DataType()

    /**
     * Класс
     */
    object Class : DataType()

    /**
     * Объект
     */
    object Object : DataType()

    /**
     * Свойство
     */
    object Property : DataType()

    /**
     * Отношение
     */
    object Relationship : DataType()

    /**
     * Строка
     */
    object String : DataType()

    /**
     * Булево значение
     */
    object Boolean : DataType()

    /**
     * Целое число
     */
    object Integer : DataType()

    /**
     * Дробное число
     */
    object Double : DataType()

    /**
     * Результат сравнения
     */
    object ComparisonResult : DataType()

    /**
     * Enum
     */
    object Enum : DataType()

    /**
     * Может ли этот тип быть преобразован в другой
     * @param to Тип, в который преобразовываем
     * @return Может ли этот тип быть преобразован в другой
     */
    fun canCast(to: DataType) = this == DecisionTreeVar && to == Object

    override fun toString() = when (this) {
        DecisionTreeVar -> "DECISION_TREE_VAR"
        Class -> "CLASS"
        Object -> "OBJECT"
        Property -> "PROPERTY"
        Relationship -> "RELATIONSHIP"
        String -> "STRING"
        Boolean -> "BOOLEAN"
        Integer -> "INTEGER"
        Double -> "DOUBLE"
        ComparisonResult -> "COMPARISON_RESULT"
        Enum -> "ENUM"
    }

    companion object {

        fun valueOf(value: kotlin.String) = when (value) {
            "DECISION_TREE_VAR" -> DecisionTreeVar
            "CLASS" -> Class
            "OBJECT" -> Object
            "PROPERTY" -> Property
            "RELATIONSHIP" -> Relationship
            "STRING" -> String
            "BOOLEAN" -> Boolean
            "INTEGER" -> Integer
            "DOUBLE" -> Double
            "COMPARISON_RESULT" -> ComparisonResult
            "ENUM" -> Enum
            else -> null
        }
    }
}